File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.java
import java.util.*;

class Solution {
    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        Map<TreeNode, TreeNode> parentMap = new HashMap<>();
        populateParents(root, null, parentMap);

        Queue<TreeNode> queue = new LinkedList<>();
        Set<TreeNode> visited = new HashSet<>();

        queue.offer(target);
        visited.add(target);

        int currentDistance = 0;

        while (!queue.isEmpty()) {
            if (currentDistance == k) {
                List<Integer> result = new ArrayList<>();
                for (TreeNode node : queue) {
                    result.add(node.val);
                }
                return result;
            }

            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode current = queue.poll();

                // Check left child
                if (current.left != null && !visited.contains(current.left)) {
                    visited.add(current.left);
                    queue.offer(current.left);
                }

                // Check right child
                if (current.right != null && !visited.contains(current.right)) {
                    visited.add(current.right);
                    queue.offer(current.right);
                }

                // Check parent node
                TreeNode parent = parentMap.get(current);
                if (parent != null && !visited.contains(parent)) {
                    visited.add(parent);
                    queue.offer(parent);
                }
            }

            currentDistance++;
        }

        return new ArrayList<>();
    }

    private void populateParents(TreeNode node, TreeNode parent, Map<TreeNode, TreeNode> parentMap) {
        if (node == null) return;
        parentMap.put(node, parent);
        populateParents(node.left, node, parentMap);
        populateParents(node.right, node, parentMap);
    }
}

File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.py
from collections import deque

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def distanceK(self, root, target, k):
        """
        :type root: TreeNode
        :type target: TreeNode
        :type k: int
        :rtype: List[int]
        """
        if not root or not target:
            return []
        
        # Step 1: Build parent map using BFS
        parent_map = {}
        queue = deque([root])
        
        while queue:
            curr = queue.popleft()
            if curr.left:
                parent_map[curr.left] = curr
                queue.append(curr.left)
            if curr.right:
                parent_map[curr.right] = curr
                queue.append(curr.right)
                
        # Step 2: BFS starting from target to find nodes at distance K
        visited = {target}
        queue = deque([target])
        current_distance = 0
        
        while queue:
            if current_distance == k:
                return [node.val for node in queue]
            
            for _ in range(len(queue)):
                curr = queue.popleft()
                
                # Check left child, right child, and parent
                neighbors = [curr.left, curr.right, parent_map.get(curr)]
                for neighbor in neighbors:
                    if neighbor and neighbor not in visited:
                        visited.add(neighbor)
                        queue.append(neighbor)
            
            current_distance += 1
            
        return []